### ISSUE TITLE
Inconsistency with TypeVars

### ISSUE THREAD

**Issue by cdce8p** [microsoft/pyright#3202]
**Describe the bug**
Came across an interesting behavior with a TypeVar bound to different scopes. Not completely sure it's a bug, however mypy does give the expected result. 

**To Reproduce**
```py
from typing import Callable, TypeVar

_TypeT = TypeVar("_TypeT", bound=type)


class DictRegistry(dict[str, _TypeT]):

    def register(self, name: str) -> Callable[[_TypeT], _TypeT]:
        def decorator(cls_type: _TypeT) -> _TypeT:
            self[name] = cls_type
            return cls_type

        return decorator


class A: ...

REGISTRY: DictRegistry[type[A]] = DictRegistry()

@REGISTRY.register("some_name")
class B(A): ...

reveal_type(B)  # should be Type[B]
reveal_type(REGISTRY.items())  # dict_items[str, type[A]]
```
```
  test.py:23:13 - information: Type of "B" is "Type[A]"
  test.py:24:13 - information: Type of "REGISTRY.items()" is "dict_items[str, Type[A]]"
```

**Expected behavior**
mypy output
```
test.py:23: note: Revealed type is "def () -> B"
test.py:24: note: Revealed type is "_collections_abc.dict_items[builtins.str*, Type[A]]"
```

**VS Code extension or command-line**
v1.1.229

**Comment by erictraut** [comment]
I think the current behavior is correct. The type of the variable `REGISTRY` is `DictRegistry[Type[A]]. A call to `REGISTRY.register("some_name")` correctly evaluates to `Type[A]` because `_TypeT` is a class-scoped TypeVar and it has been specialized to `Type[A]`.

**Comment by cdce8p** [comment]
> I think the current behavior is correct. The type of the variable `REGISTRY` is `DictRegistry[Type[A]]. A call to `REGISTRY.register("some_name")`correctly evaluates to`Type[A]`because`_TypeT`is a class-scoped TypeVar and it has been specialized to`Type[A]`.

Thought that this might be the case. Although in the example, it would make sense to assign them different. I.e. start from the inside and only use the outer scope if the TypeVar can't be resolved independently.

In the end, the decorator itself doesn't change change the type of the class after all.

**Comment by cdce8p** [comment]
Just checked the output of pyre - it's similar to the one from mypy
```
test:23:0 Revealed type [-1]: Revealed type for `B` is `typing.Type[B]`.
test:24:0 Revealed type [-1]: Revealed type for `REGISTRY.items()` is `_dict_items[str, typing.Type[A]]`.
```

**Comment by erictraut** [comment]
I'm quite certain that pyright is correct here given the scoping rules for TypeVars that are spelled out in PEP 484. The TypeVar `_TypeT` within the inner scope of `register` is defined by the outer class `DictRegistry`, and you have specialized that TypeVar with the type argument `Type[A]`. That means the method `register` is specialized to `def register(self, name: str) -> Callable[[Type[A]], Type[A]]. These are no longer type variables, so there's nothing to "solve" when the callable is invoked.

**Comment by cdce8p** [comment]
My knowledge here is limited and my previous assumption how mypy works was probably wrong. An interesting case is how mypy handles this example
```py
@REGISTRY.register("other")  # error: Argument 1 has incompatible type "Type[C]"; expected "Type[A]"
class C: ...
```

It suggests the initial specialization `Type[A]` still applies but instead of replacing `_TypeT` completely, it's added as an additional constraint. Thus preserving the TypeVar function of `_TypeT` in `register`.

--
Is there another way to express this code? The only thing I could think of which would come close is something like this here
```py
_TypeT = TypeVar("_TypeT", bound=type)
_S = TypeVar("_S", bound=_TypeT)

class DictRegistry(dict[str, _TypeT]):

    def register(self, name: str) -> Callable[[_S], _S]:
        def decorator(cls_type: _S) -> _S:
            self[name] = cls_type
            return cls_type

        return decorator
```
Although, the bound type cannot be generic.

**Comment by erictraut** [comment]
No, I don't see a good way to express this with the Python type system. Perhaps someone in the python/typing discussion forum would have some ideas.
